### ISSUE TITLE
Buffer overflow in `insert_many()`

### ISSUE THREAD

**Issue by Qwaz** [servo/rust-smallvec#252]
Hello fellow Rustacean,
we (Rust group @sslab-gatech) found a memory-safety/soundness issue in this crate while scanning Rust code on crates.io for potential vulnerabilities.

## Issue Description

https://github.com/servo/rust-smallvec/blob/9cf11764363d0a349406db752b3315168409562f/src/lib.rs#L1009-L1070

`insert_many()` overflows the buffer when an iterator yields more items than the lower bound of `size_hint()`.

The problem is in line 1044. `reserve(n)` reserves capacity for `n` more elements to be inserted. This is done by comparing the length and the capacity. Since the length of the buffer is set to 0 in line 1032, line 1044 will be always no-op and the following code will overflow the buffer.

## Reproduction

Below is an example program that exhibits undefined behavior using safe APIs of `smallvec`.

```rust
#![forbid(unsafe_code)]

use smallvec::SmallVec;

fn main() {
    let mut v: SmallVec<[u8; 0]> = SmallVec::new();

    // Spill on heap
    v.push(123);

    // Allocate string on heap
    let s = String::from("Hello!");
    println!("{}", s);

    // Prepare an iterator with small lower bound
    let mut iter = (0u8..=255).filter(|n| n % 2 == 0);
    assert_eq!(iter.size_hint().0, 0);

    // Triggering the bug
    v.insert_many(0, iter);

    // Uh oh, heap overflow made smallvec and string to overlap
    assert!(v.as_ptr_range().contains(&s.as_ptr()));

    // String is corrupted
    println!("{}", s);
}
```

Output:
```
Hello!
@BDFHJ
double free or corruption (out)

Terminated with signal 6 (SIGABRT)
```

### Tested Environment

* Crate: smallvec
* Version: 1.6.0
* OS: Ubuntu 20.04.1 LTS
* Rustc version: rustc 1.48.0 (7eac88abb 2020-11-16)

**Comment by mbrubeck** [comment]
Thanks for the report!  I have submitted a fix in #253.  I can submit a RustSec advisory after the fixed version is released, or if you would like to write the advisory, that would be fine too.

**Comment by mbrubeck** [comment]
I believe this bug was introduced in commit 8ddf61330d73bd1b33ed01e03f2bf0b8aaba8d11 (PR #213), so it affects version 1.4.2 through 1.6.0.

**Comment by Qwaz** [comment]
I believe the bug is older than that. Passing 0 as `index` will call `self.set_len(0)` and trigger the same bug.

**Comment by mbrubeck** [comment]
You're right.  This was introduced in #103, commit 26b249075930b46cfafc70b1d18fd0cb35fd2314, so it affects version 0.6.3 through 1.6.0.

**Comment by Qwaz** [comment]
I agree that that commit introduced the bug. Thank you for the quick fix!
Regarding the RustSec advisory, I think it's better to be written by the crate maintainer if possible.

**Comment by mbrubeck** [comment]
Fixed in smallvec 0.6.14 and 1.6.1.

RustSec advisory submitted in https://github.com/RustSec/advisory-db/pull/552.
